`test_destroy_unverified_incremental` is broken on XFS on Ununtu 22.04

```
elastio-snap: 
kernel: 5.15.0-35-generic
filesystem: xfs
gcc: 11.2.0-19ubuntu1)
bash: 5.1.16(1)-release
python: Python 3.10.4
test_destroy_active_incremental (test_destroy.TestDestroy) ... ok
test_destroy_active_snapshot (test_destroy.TestDestroy) ... ok
test_destroy_dormant_incremental (test_destroy.TestDestroy) ... ok
test_destroy_dormant_snapshot (test_destroy.TestDestroy) ... ok
test_destroy_nonexistent_device (test_destroy.TestDestroy) ... ok
test_destroy_unverified_incremental (test_destroy.TestDestroy) ... umount: /tmp/elastio-snap_798: target is busy.
ERROR
test_destroy_unverified_snapshot (test_destroy.TestDestroy) ... ok
test_setup_2_volumes (test_setup.TestSetup) ... ok
test_setup_already_tracked_volume (test_setup.TestSetup) ... ok
test_setup_cow_file_path_is_dir (test_setup.TestSetup) ... ok
test_setup_invalid_minor (test_setup.TestSetup) ... ok
test_setup_readonly_volume (test_setup.TestSetup) ... ok
test_setup_unmounted_volume (test_setup.TestSetup) ... ok
test_setup_volume (test_setup.TestSetup) ... ok
test_setup_volume_path_is_dir (test_setup.TestSetup) ... ok
test_modify_origin (test_snapshot.TestSnapshot) ... ok
test_next_available_minor (test_snapshot.TestSnapshot) ... ok
test_track_writes (test_snapshot.TestSnapshot) ... skipped 'Broken on XFS, due to ignored os.sync and due to #63.'
test_transition_active_incremental (test_transition_incremental.TestTransitionToIncremental) ... ok
test_transition_active_snapshot (test_transition_incremental.TestTransitionToIncremental) ... ok
test_transition_fs_sync_cow_full (test_transition_incremental.TestTransitionToIncremental) ... ok
test_transition_mod_sync_cow_full (test_transition_incremental.TestTransitionToIncremental) ... skipped 'Kernel flushed before module'
test_transition_nonexistent_snapshot (test_transition_incremental.TestTransitionToIncremental) ... ok
======================================================================
ERROR: test_destroy_unverified_incremental (test_destroy.TestDestroy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/elastio/elastio-snap/tests/test_destroy.py", line 76, in test_destroy_unverified_incremental
    util.unmount(self.mount)
  File "/home/elastio/elastio-snap/tests/util.py", line 23, in unmount
    subprocess.check_call(cmd, timeout=10)
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['umount', '/tmp/elastio-snap_798']' returned non-zero exit status 32.
----------------------------------------------------------------------
Ran 23 tests in 7.483s
FAILED (errors=1, skipped=2)
Error: Process completed with exit code 1.
```

The same "target is  busy" error is happening on `test_destroy_unverified_snapshot`.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -98,18 +98,22 @@ jobs:
         working-directory: ${{env.BOX_DIR}}
 
       - name: Install LVM and RAID tools
+        # Debian 8 is old even EOLed and it's hard to install something on it. We are keepend it just to avoid
+        # regression bugs on the old 3.16 kernel. All LVM/RAID tests are switched off on Debian 8 in the file below.
         if: "${{ matrix.distro != 'debian8' }}"
         run: |
           vagrant ssh ${{env.INSTANCE_NAME}} -c '
             if $(which apt-get >/dev/null 2>&1); then
-              sudo apt-get install -y lvm2 mdadm;
+              sudo apt-get install -y lvm2 mdadm
             else
               # Fedora has rather weak mirrors. But we do not want to have failing builds because of this.
+              set +e
               for i in {1..5}; do
                 sudo yum install -y lvm2 mdadm && break
                 echo "Failed to install LVM and RAID packages. Retrying..."
                 sleep 5
               done
+              set -e
               mdadm -V
             fi
           '
@@ -133,11 +137,10 @@ jobs:
           vagrant ssh ${{env.INSTANCE_NAME}} -c 'sudo make uninstall'
         working-directory: ${{env.BOX_DIR}}
 
-      # FIXME: enable tests on XFS for all distros.
-      # See https://github.com/elastio/elastio-snap/issues/139 about Debian 8
-      # See https://github.com/elastio/elastio-snap/issues/138 about Ubuntu 22.04
+      # FIXME: Enable tests on XFS for Debian 8. See https://github.com/elastio/elastio-snap/issues/139
+      # If we want to do that for EOLed distro.
       - name: Ignore XFS tests
-        if: "${{ matrix.distro == 'debian8' || matrix.distro == 'ubuntu2204' }}"
+        if: "${{ matrix.distro == 'debian8' }}"
         run:
           echo FS="ext2 ext3 ext4" >> $GITHUB_ENV
 
@@ -158,8 +161,11 @@ jobs:
         timeout-minutes: 1
 
       - name: Run tests on LVM (loop device)
-        # FIXME: Enable tests on all distros after resolution of #147, #143, #138
-        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian11' && matrix.distro != 'ubuntu2204' && matrix.distro != 'fedora32' }}"
+        # Debian 8 has reached EOL and has no lvm2 tools installed.
+        # But in theory we can enable tests on Debian 8 after resolution of #139.
+        # FIXME: - Enable tests on Ubuntu 20.04 after fix of #149;
+        #        - Enable tests on Ubuntu 22.04 arm64 after fix of #150.
+        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'ubuntu2004' && !(matrix.distro == 'ubuntu2204' && matrix.arch == 'arm64') }}"
         run: |
           for fs in ${FS[*]}; do
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -f $fs --lvm"
@@ -176,8 +182,9 @@ jobs:
         timeout-minutes: 1
 
       - name: Run tests on RAID (loop device)
-        # FIXME: Enable tests on all distros after resolution of #146, #147, #143, #138
-        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian9' && matrix.distro != 'debian11' && matrix.distro != 'ubuntu2204' && matrix.distro != 'fedora32' }}"
+        # FIXME: Enable tests on Debian 9 after resolution of #146.
+        # And Debian 8 has reached EOL and has no mdadm tools installed.
+        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian9' }}"
         run: |
           for fs in ${FS[*]}; do
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -f $fs --raid"
@@ -232,8 +239,11 @@ jobs:
         timeout-minutes: 1
 
       - name: Run tests on LVM (qcow2 disks)
-        # FIXME: Enable tests on all distros after resolution of #147, #143, #138
-        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian11' && matrix.distro != 'ubuntu2204' && matrix.distro != 'fedora32' }}"
+        # Debian 8 has reached EOL and has no lvm2 tools installed.
+        # But in theory we can enable tests on Debian 8 after resolution of #139.
+        # FIXME: - Enable tests on Ubuntu 20.04 after fix of #149;
+        #        - Enable tests on Ubuntu 22.04 arm64 after fix of #150.
+        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'ubuntu2004' && !(matrix.distro == 'ubuntu2204' && matrix.arch == 'arm64') }}"
         run: |
           for fs in ${FS[*]}; do
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb1 -d /dev/vdc1 -f $fs --lvm"
@@ -248,8 +258,9 @@ jobs:
         timeout-minutes: 1
 
       - name: Run tests on RAID (qcow2 disks)
-        # FIXME: Enable tests on all distros after resolution of #146, #147, #143, #138
-        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian9' && matrix.distro != 'debian11' && matrix.distro != 'ubuntu2204' && matrix.distro != 'fedora32' }}"
+        # FIXME: Enable tests on Debian 9 after resolution of #146.
+        # And Debian 8 has reached EOL and has no mdadm tools installed.
+        if: "${{ matrix.distro != 'debian8' && matrix.distro != 'debian9' }}"
         run: |
           for fs in ${FS[*]}; do
             vagrant ssh ${{env.INSTANCE_NAME}} -c "cd tests && sudo ./elio-test.sh -d /dev/vdb1 -d /dev/vdc1 -f $fs --raid"

--- a/src/configure-tests/feature-tests/bvec_iter_all.c
+++ b/src/configure-tests/feature-tests/bvec_iter_all.c
@@ -1,0 +1,17 @@
+// SPDX-License-Identifier: GPL-2.0-only
+
+/*
+ * Copyright (C) 2015 Datto Inc.
+ * Additional contributions by Elastio Software, Inc are Copyright (C) 2020 Elastio Software Inc.
+ */
+
+#include "includes.h"
+MODULE_LICENSE("GPL");
+
+static inline void dummy(void){
+	struct bvec_iter_all iter;
+
+	iter.idx = 0;
+	iter.done = 0;
+	memset(&iter.bv, 0, sizeof(iter.bv));
+}

--- a/src/configure-tests/feature-tests/bvec_iter_all.c
+++ b/src/configure-tests/feature-tests/bvec_iter_all.c
@@ -1,8 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 /*
- * Copyright (C) 2015 Datto Inc.
- * Additional contributions by Elastio Software, Inc are Copyright (C) 2020 Elastio Software Inc.
+ * Copyright (C) 2020 Elastio Software Inc.
  */
 
 #include "includes.h"

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -2991,6 +2991,12 @@ static int snap_handle_write_bio(const struct snap_device *dev, struct bio *bio)
 	char *data;
 	sector_t start_block, end_block = SECTOR_TO_BLOCK(bio_sector(bio));
 
+	/*
+	 * Previously we iterated using bio_for_each_segment(), which
+	 * caused problems in case if our bio was split by the system.
+	 * It is replaced with bio_for_each_segment_all() as we own the
+	 * bio and can guarantee that we have access to its bvecs
+	 */
 #ifdef HAVE_BVEC_ITER_ALL
 	struct bvec_iter_all iter;
 	struct bio_vec *bvec;

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -816,12 +816,18 @@ __kernel_ulong_t si_mem_available(void)
 
 #ifndef HAVE_BIO_FREE_PAGES
 static void bio_free_pages(struct bio *bio){
-	bio_iter_t iter;
-	bio_iter_bvec_t bvec;
 	struct page *bv_page;
 
-	bio_for_each_segment(bvec, bio, iter) {
-		bv_page = bio_iter_page(bio, iter);
+#ifdef HAVE_BVEC_ITER_ALL
+	struct bvec_iter_all iter;
+	struct bio_vec *bvec;
+	bio_for_each_segment_all(bvec, bio, iter) {
+#else
+	int i = 0;
+	struct bio_vec *bvec;
+	bio_for_each_segment_all(bvec, bio, i) {
+#endif
+		bv_page = bvec->bv_page;
 		if (bv_page) {
 			__free_page(bv_page);
 		}
@@ -3200,7 +3206,6 @@ static void __on_bio_read_complete(struct bio *bio, int err){
 	for(map = tp->bio_sects.head; map != NULL && map->bio != NULL; map = map->next) {
 		if(bio == map->bio){
 			bio_sector(bio) = map->sect - dev->sd_sect_off;
-			bio_size(bio) = map->size;
 			bio_idx(bio) = 0;
 			break;
 		}

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -2982,32 +2982,29 @@ out:
 
 static int snap_handle_write_bio(const struct snap_device *dev, struct bio *bio){
 	int ret;
-	bio_iter_t iter;
-	bio_iter_bvec_t bvec;
+	struct bvec_iter_all iter;
+	struct bio_vec *bvec;
 	char *data;
 	sector_t start_block, end_block = SECTOR_TO_BLOCK(bio_sector(bio));
 
 	//iterate through the bio and handle each segment (which is guaranteed to be block aligned)
-	bio_for_each_segment(bvec, bio, iter){
+	bio_for_each_segment_all(bvec, bio, iter){
 		//find the start and end block
 		start_block = end_block;
-		end_block = start_block + (bio_iter_len(bio, iter) / COW_BLOCK_SIZE);
-
+		end_block = start_block + (bvec->bv_len / COW_BLOCK_SIZE);
 		//map the page into kernel space
-		data = kmap(bio_iter_page(bio, iter));
-
+		data = kmap(bvec->bv_page);
 		//loop through the blocks in the page
 		for(; start_block < end_block; start_block++){
 			//pas the block to the cow manager to be handled
 			ret = cow_write_current(dev->sd_cow, start_block, data);
 			if(ret){
-				kunmap(bio_iter_page(bio, iter));
+				kunmap(bvec->bv_page);
 				goto error;
 			}
 		}
-
 		//unmap the page
-		kunmap(bio_iter_page(bio, iter));
+		kunmap(bvec->bv_page);
 	}
 
 	return 0;

--- a/src/elastio-snap.c
+++ b/src/elastio-snap.c
@@ -2982,13 +2982,19 @@ out:
 
 static int snap_handle_write_bio(const struct snap_device *dev, struct bio *bio){
 	int ret;
-	struct bvec_iter_all iter;
-	struct bio_vec *bvec;
 	char *data;
 	sector_t start_block, end_block = SECTOR_TO_BLOCK(bio_sector(bio));
 
+#ifdef HAVE_BVEC_ITER_ALL
+	struct bvec_iter_all iter;
+	struct bio_vec *bvec;
 	//iterate through the bio and handle each segment (which is guaranteed to be block aligned)
-	bio_for_each_segment_all(bvec, bio, iter){
+	bio_for_each_segment_all(bvec, bio, iter) {
+#else
+	int i = 0;
+	struct bio_vec *bvec;
+	bio_for_each_segment_all(bvec, bio, i) {
+#endif
 		//find the start and end block
 		start_block = end_block;
 		end_block = start_block + (bvec->bv_len / COW_BLOCK_SIZE);

--- a/tests/test_destroy.py
+++ b/tests/test_destroy.py
@@ -40,9 +40,6 @@ class TestDestroy(DeviceTestCase):
         self.assertFalse(os.path.exists(self.snap_device))
         self.assertIsNone(elastio_snap.info(self.minor))
 
-    @unittest.skipIf(int(platform.release().split(".")[0]) == 5
-                     and int(platform.release().split(".")[1]) == 10,
-                     "Broken on Debian 11 with kernel 5.10 (see #147)")
     def test_destroy_dormant_snapshot(self):
         self.assertEqual(elastio_snap.setup(self.minor, self.device, self.cow_full_path), 0)
 
@@ -54,9 +51,6 @@ class TestDestroy(DeviceTestCase):
         self.assertFalse(os.path.exists(self.snap_device))
         self.assertIsNone(elastio_snap.info(self.minor))
 
-    @unittest.skipIf(int(platform.release().split(".")[0]) == 5
-                     and int(platform.release().split(".")[1]) == 10,
-                     "Broken on Debian 11 with kernel 5.10 (see #147)")
     def test_destroy_dormant_incremental(self):
         self.assertEqual(elastio_snap.setup(self.minor, self.device, self.cow_full_path), 0)
         self.assertEqual(elastio_snap.transition_to_incremental(self.minor), 0)

--- a/tests/test_transition_incremental.py
+++ b/tests/test_transition_incremental.py
@@ -38,10 +38,6 @@ class TestTransitionToIncremental(DeviceTestCase):
         self.assertEqual(elastio_snap.transition_to_incremental(self.minor), 0)
         self.assertEqual(elastio_snap.transition_to_incremental(self.minor), errno.EINVAL)
 
-    @unittest.skipIf((os.getenv('LVM') or os.getenv('RAID'))
-                     and os.getenv('TEST_FS') == "xfs"
-                     and int(platform.release().split(".")[0]) == 5,
-                     "Broken on LVM/raid with XFS, kernels 5.9+ (see #143) and kernel 5.8 (see #145)")
     def test_transition_fs_sync_cow_full(self):
         scratch = "{}/scratch".format(self.mount)
         falloc = 50
@@ -67,10 +63,6 @@ class TestTransitionToIncremental(DeviceTestCase):
         self.assertEqual(snapdev["error"], -errno.EFBIG)
         self.assertEqual(snapdev["state"], 3)
 
-    @unittest.skipIf((os.getenv('LVM') or os.getenv('RAID'))
-                     and os.getenv('TEST_FS') == "xfs"
-                     and int(platform.release().split(".")[0]) == 5,
-                     "Broken on LVM/raid with XFS, kernels 5.9+ (see #143) and kernel 5.8 (see #145)")
     def test_transition_mod_sync_cow_full(self):
         scratch = "{}/scratch".format(self.mount)
         falloc = 50

--- a/tests/util.py
+++ b/tests/util.py
@@ -8,6 +8,7 @@
 
 import hashlib
 import subprocess
+import sys
 import time
 
 
@@ -19,9 +20,27 @@ def mount(device, path, opts=None):
     subprocess.check_call(cmd, timeout=10)
 
 
-def unmount(path):
+def unmount(path, retry_on_dev_busy=True):
     cmd = ["umount", path]
-    subprocess.check_call(cmd, timeout=10)
+    # subprocess.run is introduced in Python 3.5
+    if not retry_on_dev_busy or sys.version_info <= (3, 5):
+        subprocess.check_call(cmd, timeout=10)
+    else:
+        # The retries on device busy error are necessary on Ubuntu 22.04, kernel 5.15
+        # for the tests test_destroy_unverified_incremental and test_destroy_unverified_snapshot.
+        # See https://github.com/elastio/elastio-snap/issues/138
+        retries = 3
+        for retry in range(retries):
+            p = subprocess.run(cmd, timeout=20, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            if p.returncode == 0:
+                break
+            elif retry + 1 < retries and "busy" not in p.stderr.decode():
+                print("Command umount " + path + " has failed (" + str(p.returncode) + "): " + p.stderr.decode())
+                raise subprocess.CalledProcessError(p.returncode, cmd, "Command failed")
+            elif retry + 1 == retries:
+                raise subprocess.CalledProcessError(p.returncode, cmd, "Command failed " + str(retries) + "times")
+            else:
+                time.sleep(1)
 
 
 def dd(ifile, ofile, count, **kwargs):
@@ -104,21 +123,21 @@ def parted_create_lvm_raid_partitions(devices, kind):
     partitions=[]
     for device in devices:
         cmd = ["wipefs", "--all", "--force", "--quiet", device]
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=30)
         cmd = ["parted", "--script", device, "mklabel gpt"]
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=30)
         cmd = ["parted", "--script", device, "mkpart '" + part_type + "' 0% 100%"]
         subprocess.check_call(cmd, timeout=30)
         cmd = ["parted", "--script", device, "set 1 " + kind + " on"]
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=30)
         cmd = ["partprobe", device]
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=30)
         settle()
         part = get_last_partition(device)
         # mdadm rarely and randomly complains on create about superblock
         # let's clean it up and do not care about return code
         cmd = ["wipefs", "--all", "--force", "--quiet", part]
-        subprocess.check_call(cmd, timeout=10)
+        subprocess.check_call(cmd, timeout=30)
         partitions.append(part)
 
     return partitions

